Compare marks/measures to performance.now mark

tollmanz opened the pull request:
When using the User Timing API plugin, only the first beacon sent will include User Timing API data. Subsequent beacons will not collect newly set User Timing API data due to comparing incorrect timestamps.

To reproduce:

1. Add code to a test page running Boomerang and the User Timing API plugin. Something like the following would suffice:

   ```
   performance.mark('pre-load-mark');

   addEventListener('load', () => {
     setTimeout(() => {
       console.log('marking');
       performance.mark('post-load-mark');
     }, 5000);
   });
   ```

2. Load the page with Boomerang, the User Timing API plugin and the sample code
3. After the `load` event fires, inspect the Network tab to see that a beacon was sent with the User Timing API data, including the 'pre-load-mark' mark. The 'post-load-mark' should not be present.
4. Fire another beacon from the console  with: `BOOMR.requestStart('test').loaded();`
5. Notice that the expected 'post-load-mark' is not present in the beacon.

On line [41 in plugins/usertiming.js](https://github.com/SOASTA/boomerang/blob/master/plugins/usertiming.js#L41), the `from` value is set to `BOOMR.now()`. This value is a millisecond timestamp derived from `Date.now()`. When subsequent beacons are fired, the mark/measure's `startTime + duration` is [compared](https://github.com/nicjansma/usertiming-compression.js/blob/master/src/usertiming-compression.js#L388) to the `from` timestamp. Since the mark/measure is using
`performance.now()` instead of `Date.now()` to set the time, the value will always be less than the `Date.now()` timestamp. Thus, the data is considered old and is removed from the compressed data.

The fix for the issue is to use `performance.now()` instead of `Date.now()` when setting the `from` value. Since this function isn't called unless the User Timing API is available, it should be safe to directly call `performance.now()` without any defensive coding.

bluesmoon commented on plugins/usertiming.js:
```diff
@@ -33,7 +33,7 @@
 		 */
 		getUserTiming: function() {
-			var timings, res, now = BOOMR.now();
+			var timings, res, now = performance.now();
```
unfortunately this is not safe to use if the site uses a UserTiming shim that does not also shim `performance.now()`

tollmanz replied:
Yeah...I see now. I was looking at https://github.com/SOASTA/boomerang/blob/master/plugins/usertiming.js#L117, but it does not shim `performance.now()`.

tollmanz replied:
@bluesmoon Any chance you can explain [this comment](https://github.com/SOASTA/boomerang/blob/master/boomerang.js#L1190-L1192)? Should `BOOMR.now()` actually be producing the correct timestamp?

bluesmoon replied:
well, `BOOMR.now()` is supposed to return a unix timestamp using the best available timing mechanism in the browser. This is different from `performance.now()` which gives you the timestamp relative to `navigationStart`. The question is, what's actually needed by this code?

bluesmoon replied:
unfortunately it looks like that particular piece of code isn't merged into master.  it should be this:

```javascript

	// If the browser supports performance.now(), swap that in for BOOMR.now
	try {
		var p = boomr.getPerformance();
		if (p &&
		    typeof p.now === "function" &&
		    /\[native code\]/.test(String(p.now)) &&		// #545 handle bogus performance.now from broken shims
		    p.timing &&
		    p.timing.navigationStart) {
			boomr.now = function() {
				return Math.round(p.now() + p.timing.navigationStart);
			};
		}
	}
	catch (ignore) {
		// empty
	}
```

bluesmoon replied:
let me see how easy it is to merge this in right now

tollmanz replied:
Thanks! That's what I thought.

My understanding is that this code is trying to mark when data from the User Timing API was last accessed. By marking that time, it knows which marks/measures to _not_ send when subsequent beacons are sent.

Let's say you do the following:

1. Set performance mark "A"
1. Send a beacon, in which performance mark "A" is sent
1. Set performance mark "B"
1. Send a second beacon, which does not include "B"

The problem is that during the sending of the first beacon, the plugin sets a `from` value that will look like `1510690156237`. When the second beacon is sent, it compares that value with the `startTime` and `duration` times in each of the marks/measures. These values look like `
112665.99`. The mark/measures values will _always_ be less than `BOOMR.now()` and therefore the values are discarded.

The plugin presumably does this because it does not want to clear marks that have already been beaconed. Thus, it tries to determine which marks were added _after_ the previous beacon was sent.

bluesmoon replied:
Sorry, looks like it's already in there but I couldn't figure out how to search:
https://github.com/SOASTA/boomerang/blob/master/boomerang.js#L1964-L1979

bluesmoon replied:
@querymetrics can you look into this?

tollmanz replied:
Seeing this in Chrome 61:

<img width="229" alt="screen shot 2017-11-14 at 2 19 50 pm" src="https://user-images.githubusercontent.com/921795/32802799-f2d439ec-c946-11e7-91e9-40e94c75422b.png">

bluesmoon replied:
right, `BOOMR.getPerformance()` just returns the `performance` object in a browser independent way.

tollmanz replied:
Ok...I think I'm finally understanding what's going on. `BOOMR.now()` _should_ be a high precise timestamp from the unix epoch. When `performance.now()` is available, that timestamp is added to `performance.timing.navigationStart` to get a more accurate timestamp than `Date.now()` provides.

If I'm understanding this correctly, then `BOOMR.now()` is the incorrect timestamp to use.

What if we did something like:

```
BOOMR.getPerformance.now() || BOOMR.now()
```

In the fallback case, we'd still have this problem, but browsers that support the User Timing API should be fine.

bluesmoon replied:
that would throw an exception if the shim doesn't exist.  I think this would work:

```javascript
var p = BOOMR.getPerformance()

// HiRes Timestamp supported
if (p && p.now) {
    now = p.now();
}
// HiRes Timestamp not supported, but RT Plugin exists and has figured out navigationStart
else if (BOOMR.plugins.RT && BOOMR.plugins.RT.navigationStart) {
    now = BOOMR.now() - BOOMR.plugins.RT.navigationStart();
}
// NavigationTiming not supported, but either BOOMR.t_lstart or BOOMR.t_start are set
// which are when boomerang started and not exactly when the page started
else {
    now = BOOMR.now() - (BOOMR.t_lstart || BOOMR.t_start);
}
```

This will do everything possible to get the right timestamp relative to navigationStart.

tollmanz replied:
This is great. I didn't know many of these things existed. I'll integrate this suggestion and see it is works.

Would it be worth moving this code to the main script to have a canonical function for `performance.now()`-esque values?

tollmanz commented:
@bluesmoon I've updated the code based on your recommendations. This uses the same general idea that you suggested, but changes the syntax slightly.

I've tested this for each of the 3 ways that the time since navigation start could be computed and I get the desired results. Obviously, there could be edge cases where the inaccurate navigation start time could cause some data to be duplicated or missed. I think this will be a somewhat rare occurrence.

tollmanz commented:
Note that I force pushed this to maintain a clean branch. You are best to pull the changes via `git pull --rebase` to avoid issues.

bluesmoon commented:
looks good to me.  I'd still like @querymetrics metrics to review this as the primary maintainer of this plugin.

querymetrics commented:
@tollmanz Thanks for this. Would you be able to add unit test(s) in `tests/page-templates/18-usertiming` that sends multiple beacons ? If not, i could create one.

tollmanz commented:
@querymetrics Can do! Being unfamiliar with the test setup, I missed that you had existing tests.

tollmanz commented:
@querymetrics I'm having a heck of a time getting `npm test` to work. I've run into Phantom JS issues, missing dependency problems and now Web Driver issues. 

So far, to run the tests, I done:

1. `npm i`
1. `bower install`
1. Installed a global version of Phantom JS and referenced it with `PHANTOMJS_BIN=/path/to/phantomjs` to run tests
1. Installed Java

Is there anything else I should be doing to get tests to run? Can you tell me what version of Node you are using to run the tests?

Thank you!

querymetrics commented:
You shouldn't need to install Phantom globally, `npm` will pull it in locally. That said, if you're using OSX Sierra or higher it probably won't work, Phantom 1.x crashes.
If you want to run your test in a "real" browser instead of running the whole suite on the command line, simply run `grunt clean build test:debug` then navigate to `http://localhost:4002/`. Once that's up, you can edit a test and it should recompile automatically.
It should work with the latest node (i'm using 8.9.1).

tollmanz commented:
> You shouldn't need to install Phantom globally, npm will pull it in locally. That said, if you're using OSX Sierra or higher it probably won't work, Phantom 1.x crashes.

This was the problem I was having and found that global Phantom helped work around it.

I'll try the other method.

tollmanz commented:
@querymetrics I've updated the commit with the requested tests. Is there a better way to send a beacon than `BOOMR.requestStart('test').loaded();`? I've tried `BOOMR.sendBeacon()`, but get a console message stating that the RT plugin is not loaded and the beacon cannot be sent.

If this is an ok way to send the beacon, then I think this should be ok.

tollmanz commented:
@querymetrics How's this looking? Anything else you need from me to get this merged in?

querymetrics commented on tests/page-templates/18-usertiming/04-usertiming-second-beacon.html:
```diff
@@ -0,0 +1,30 @@
+  }
+
+  BOOMR_test.init({
+    testAfterOnBeacon: true,
```
Change to `testAfterOnBeacon: 2,`. This will make the test to wait for 2 beacons before running checks otherwise it will be timing sensitive.

tollmanz replied:
Cool! I wasn't sure what that variable did.

querymetrics commented on tests/page-templates/18-usertiming/04-usertiming-second-beacon.html:
```diff
@@ -0,0 +1,30 @@
+      window.performance.mark("post-load-mark-start");
+      window.performance.mark("post-load-mark-end");
+      window.performance.measure("post-load-measure", "post-load-mark-start", "post-load-mark-end");
+      BOOMR.requestStart('test').loaded();
```
Could simplify this to `BOOMR.responseEnd("test");`

tollmanz replied:
Great! I did not know you could do that.

querymetrics commented:
There's a few tab vs space occurrences, `grunt lint` will point them out for you.

tollmanz commented:
@querymetrics Thanks for the review! I've added the new tests. I added a new HTML page and JS test. This is all very redundant, but I'm not sure if there is a way to set a variable for the HTML page that could allow me to turn `performance.now()` on and off. Please let me know if there is a better way. I could not find any such examples in the test suite.

Additionally, I should also test the condition of when `BOOMR.plugins.RT && BOOMR.plugins.RT.navigationStart` returns `false`; however, I'm not sure how to simulate this. Do you have any pointers?

Thanks for all of your help!

querymetrics commented on tests/page-templates/18-usertiming/05-usertiming-second-beacon-no-performance-now.html:
```diff
@@ -0,0 +1,32 @@
+<%= header %>
+<%= boomerangSnippet %>
+<script>
+	window.performance.now = undefined;
```
we might need a `if (window.performance)` here just in case

querymetrics commented on .idea/codeStyleSettings.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Could you remove this file

tollmanz replied:
Ugh...apologies for letting that creep in.

querymetrics commented on tests/page-templates/18-usertiming/05-usertiming-second-beacon-no-performance-now.html:
```diff
@@ -0,0 +1,32 @@
+	}
+
+	BOOMR_test.init({
+		testAfterOnBeacon: 2,
```
Now that i think of it, we should do `BOOMR_test.isUserTimingSupported() ? 2 : 1` here. The 2nd beacon won't get sent by `sendPostLoadMark` if UT is not supported. Same for other test page.

querymetrics commented:
The 2nd test looks good, don't worry about the redundancy.
There might not be an easy way to test for missing `BOOMR.plugins.RT`, our test build includes it.

tollmanz commented:
@querymetrics Latest round of feedback has been addressed. Thanks again.

querymetrics commented:
@tollmanz Thanks for finding & fixing this!

tollmanz commented:
@querymetrics Thank you for all of the help! I'd be happy to rebase into a single commit to keep the commit history looking good. Let me know if you would like me to do that.

tollmanz commented:
@querymetrics I went ahead and squashed the commits into one and signed it per @bluesmoon's suggestion.
